### PR TITLE
opt: reuse catalog across queries

### PR DESCRIPTION
**Pull request by RaduBerinde**
This change moves the planning context to the planner and reuses a
single instance of the optCatalog. The optCatalog caches data sources,
keyed by `*ImmutableTableDescriptor`. This saves the creation of new
data sources each time we plan a query, and even when we check a
cached memo for staleness.

Benchmarks (with GOMAXPROCS=1):
```
QueryCache/small/clients-1/simple/cache-off         227µs ± 2%   226µs ± 5%    ~     (p=0.971 n=10+10)
QueryCache/small/clients-1/simple/cache-on          170µs ± 2%   168µs ± 3%    ~      (p=0.133 n=9+10)
QueryCache/small/clients-1/prepare-once/cache-off   182µs ± 2%   178µs ± 1%  -1.90%   (p=0.010 n=10+9)
QueryCache/small/clients-1/prepare-once/cache-on    181µs ± 2%   178µs ± 2%  -1.55%   (p=0.008 n=9+10)
QueryCache/small/clients-1/prepare-each/cache-off   304µs ± 3%   298µs ± 2%  -1.84%  (p=0.002 n=10+10)
QueryCache/small/clients-1/prepare-each/cache-on    264µs ± 4%   258µs ± 4%  -2.47%  (p=0.029 n=10+10)
```

Release note: None

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Ah.. this breaks cache invalidation when statistics are created. Let me work on it some more.

**Comment by RaduBerinde** [comment]
Updated. I'm checking if the `[]*TableStatistics` slice is the same (same underlying array), which should be very fast but feels a bit hacky.

**Comment by RaduBerinde** [comment]
image

**Review by justinj — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-:-L_DTqBy-zbDLqm_gnlQ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @justinj)

**Review by andy-kimball — COMMENTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-:-L_DUsIt1lmd7cGPEoVZ:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @RaduBerinde)

___
*[pkg/sql/opt_catalog.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-L_DR-Xp4O9fPnZ9dLY1:-L_DR-Xq2Mv2qFA2XmdC:b-rbcs6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a47cb12e8fb7d6699aaeec5aa46bdc9a92c2b37f/pkg/sql/opt_catalog.go#L204)):*
> ```Go
> 	if ds, ok := oc.dataSources[desc]; ok {
> 		if t, ok := ds.(*optTable); ok {
> 			// Fast check to verify that the statistics haven't changed: we check the
> ```

There's going to be a merge conflict between this change and the latency change, since zones act like statistics (i.e. they have to be checked for changes as well). I think you should break out this `IsStale` code into a separate function for clarity, since this logic may continue to get more complex.
___
*[pkg/sql/plan_opt.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-L_DS2keEL4oLUalWL4o:-L_DS2keEL4oLUalWL4p:bmkkb4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a47cb12e8fb7d6699aaeec5aa46bdc9a92c2b37f/pkg/sql/plan_opt.go#L198)):*
> ```Go
> 	p *planner
> 
> 	// catalog is initialized once, and reset for each query.
> ```

I'd add, `This allows the catalog objects to be reused across queries in the same session.`

**Review by andy-kimball — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-:-L_DUwXG7d-Vso9RxxH4:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @RaduBerinde)

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/opt_catalog.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-L_DR-Xp4O9fPnZ9dLY1:-L_DWdRaEBYEBsnNKu_g:btkuxlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a47cb12e8fb7d6699aaeec5aa46bdc9a92c2b37f/pkg/sql/opt_catalog.go#L204)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

There's going to be a merge conflict between this change and the latency change, since zones act like statistics (i.e. they have to be checked for changes as well). I think you should break out this `IsStale` code into a separate function for clarity, since this logic may continue to get more complex.
</blockquote></details>

Sounds good, I'm going to wait for that to merge and then do that.

**Review by RaduBerinde — COMMENTED** [review]
Integrated with the zone config changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-:-L_EiEEr12VfbsyUVzDJ:brw9fxy)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @justinj)

___
*[pkg/sql/opt_catalog.go, line 204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-L_DR-Xp4O9fPnZ9dLY1:-L_EiaT13dyIVOE1R6b3:b-lmaytl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a47cb12e8fb7d6699aaeec5aa46bdc9a92c2b37f/pkg/sql/opt_catalog.go#L204)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sounds good, I'm going to wait for that to merge and then do that.
</blockquote></details>

Done. Refactored this code a bit.
___
*[pkg/sql/plan_opt.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-L_DS2keEL4oLUalWL4o:-L_EgykLEUM9-W-zAMZr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a47cb12e8fb7d6699aaeec5aa46bdc9a92c2b37f/pkg/sql/plan_opt.go#L198)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd add, `This allows the catalog objects to be reused across queries in the same session.`
</blockquote></details>

Done.

**Comment by RaduBerinde** [comment]
I'd like another look before I merge (touched a bunch of things related to the zone config).

**Review by andy-kimball — APPROVED** [review]
:lgtm:, thanks for doing the zone config work.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35270#-:-L_JHfQl6VE9r3iVBpgN:b-6x971q)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @justinj)

**Comment by RaduBerinde** [comment]
TFTRs!

bors r+

**Pull request merged**
